Iter1 → master

dtroode opened the pull request.

dtroode committed:
feat: iteration 1

dtroode committed:
test: add tests

dtroode committed:
ci: remove go tool

dtroode committed:
build: down go versions for ci support

dtroode committed:
test: use custom type for context key

dtroode committed:
test: close response body

dtroode committed:
test: fix resp body closing

dtroode committed:
chore: add ability to stop workerpool and pass duration to job

dtroode committed:
test: add job tests

dtroode committed:
docs: update swagger

bulbetski commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,90 @@
+
+	store, err := postgres.NewStorage(cfg.DatabaseDSN)
+	if err != nil {
+		log.Error("failed to start db conn", "error", err)
+		os.Exit(1)
```
можно заменить на log.Fatalf

dtroode replied:
я уже к этому времени ввожу свой кастомный логгер, у него нет fatalf. не хочу придумывать новое имя для переменной моего логгера чтоб дефолтный глобальный был доступен для вызова. все равно log.fatal логирует и вызывает os.Exit(1) точно так же

bulbetski commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,90 @@
+
+	log := logger.NewLog(cfg.LogLevel)
+
+	log.Info("config: ", "config", cfg)
```
если для отладки, то ок. В проде не надо.
Тем более логируешь секреты типа JWTSecretKey

bulbetski commented on database/migrations/20250517134728_create_table_users.sql:
```diff
@@ -0,0 +1,15 @@
+    login varchar(64) NOT NULL UNIQUE,
+    password varchar(64) NOT NULL,
+    created_at timestamptz NOT NULL DEFAULT now(),
+    balance real DEFAULT 0
```
надо взять за привычку все "деньги" надо хранить в интах. Это надо делать из-за арифметики флоатов и кривого окруления. На больших объемах будем терять/получать лишнее

bulbetski commented on internal/.DS_Store:
игнорим бинарник

bulbetski commented on internal/application/service/job.go:
какая-то жесткая вложенность неприятная, надо как-то более плоско сделать, разбить на хелперы мб.

bulbetski replied:
Может что-то подобное
```
func (s *Service) checkOrderJob(orderID uuid.UUID, orderNumber string, interval time.Duration) func(ctx context.Context) (any, error) {
	return func(ctx context.Context) (any, error) {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		currentStatus := model.AccrualOrderStatusRegistered

		for {
			select {
			case <-ticker.C:
				order, err := s.accrualAdapter.GetOrder(ctx, orderNumber)
				if err != nil {
					if isRetryableError(err) {
						continue
					}
					return nil, fmt.Errorf("failed to get order info: %w", err)
				}

				switch order.Status {
				case model.AccrualOrderStatusProcessing:
					if currentStatus != order.Status {
						if err := s.updateOrderStatus(ctx, orderID, model.OrderStatusProcessing); err != nil {
							return nil, err
						}
						currentStatus = order.Status
					}
					continue

				case model.AccrualOrderStatusInvalid:
					return s.finalizeInvalidOrder(ctx, orderID)

				case model.AccrualOrderStatusProcessed:
					return s.finalizeProcessedOrder(ctx, orderID, order.Accrual)

				default:
					// Status is REGISTERED or unknown — continue polling
					continue
				}

			case <-ctx.Done():
				return nil, fmt.Errorf("order job canceled: %w", ctx.Err())
			}
		}
	}
}
```

```
func isRetryableError(err error) bool {
	return errors.Is(err, application.ErrAccrualTooManyRequests) ||
		errors.Is(err, application.ErrAccrualOrderNotRegistered)
}

func (s *Service) updateOrderStatus(ctx context.Context, orderID uuid.UUID, status model.OrderStatus) error {
	params := &storage.SetOrderStatus{
		ID:     orderID,
		Status: status,
	}
	if _, err := s.storage.SetOrderStatus(ctx, params); err != nil {
		return fmt.Errorf("failed to update order status: %w", err)
	}
	return nil
}

func (s *Service) finalizeInvalidOrder(ctx context.Context, orderID uuid.UUID) (any, error) {
	params := &storage.SetOrderStatus{
		ID:     orderID,
		Status: model.OrderStatusInvalid,
	}
	order, err := s.storage.SetOrderStatus(ctx, params)
	if err != nil {
		return nil, fmt.Errorf("failed to finalize invalid order: %w", err)
	}
	return order, nil
}

func (s *Service) finalizeProcessedOrder(ctx context.Context, orderID uuid.UUID, accrual int) (any, error) {
	params := &storage.SetOrderStatusAndAccrual{
		ID:      orderID,
		Status:  model.OrderStatusProcessed,
		Accrual: accrual,
	}
	order, err := s.storage.SetOrderStatusAndAccrual(ctx, params)
	if err != nil {
		return nil, fmt.Errorf("failed to finalize processed order: %w", err)
	}
	return order, nil
}
```

bulbetski replied:
всё равно неприятно

dtroode replied:
ну хотя бы так, взял этот вариант

bulbetski commented on internal/application/service/job.go:
```diff
@@ -0,0 +1,81 @@
+
+func (s *Service) checkOrderJob(orderID uuid.UUID, orderNumber string, duration time.Duration) func(ctx context.Context) (any, error) {
+	return func(ctx context.Context) (any, error) {
+		c := time.Tick(duration)
```
лучше использовать time.NewTicker

dtroode replied:
в доке к `time.Tick()`
> Tick is a convenience wrapper for [NewTicker] providing access to the ticking channel only. Unlike NewTicker, Tick will return nil if d <= 0.

мне ничего кроме канала не нужно, поэтому думаю лучше так оставить

bulbetski replied:
ок, не знал)

bulbetski commented on internal/application/service/job.go:
```diff
@@ -0,0 +1,81 @@
+					return updatedOrder, nil
+				}
+			case <-ctx.Done():
+				return nil, fmt.Errorf("failed to check order, context deadline exceeded")
```
Сообщение не соответствует действительности — может быть не только "deadline exceeded". Лучше использовать err := ctx.Err() и вернуть его обёрткой.

bulbetski commented on internal/application/service/service.go:
```diff
@@ -0,0 +1,270 @@
+	hasher         Hasher
+	tokenManager   TokenManager
+	accrualAdapter AccrualAdapter
+	pool           WorkerPool
```
Не уверен насчет интерфейса на воркер пул. Как будто и структуру можно

dtroode replied:
тестировать удобнее

bulbetski commented on internal/application/service/service.go:
```diff
@@ -0,0 +1,270 @@
+
+func (s *Service) RegisterUser(ctx context.Context, params *request.RegisterUser) (string, error) {
+	_, err := s.storage.GetUserByLogin(ctx, params.Login)
+	if err != nil {
+		if !errors.Is(err, application.ErrNotFound) {
+			return "", fmt.Errorf("failed to check user with login: %w", err)
+		}
+	} else {
+		return "", application.ErrConflict
+	}
```
тут получается, что в случае nil ошибки ты возвращаешь application.ErrConflict

bulbetski commented on internal/application/service/service.go:
```diff
@@ -0,0 +1,270 @@
+	return token, nil
+}
+
+func (s *Service) checkByLuna(number string) error {
```
luna => luhn

bulbetski commented on internal/application/storage/storage.go:
```diff
@@ -0,0 +1,55 @@
+	GetUserOrdersNewestFirst(ctx context.Context, userID uuid.UUID) ([]*model.Order, error)
+}
+
+type SetOrderAccrual struct {
+	ID      uuid.UUID
+	Accrual float32
+}
+
+type SetOrderStatus struct {
+	ID     uuid.UUID
+	Status model.OrderStatus
+}
+
+type SetOrderStatusAndAccrual struct {
+	ID      uuid.UUID
+	Status  model.OrderStatus
+	Accrual float32
+}
+
+type SetUserBalance struct {
+	ID      uuid.UUID
+	Balance float32
+}
+
+type IncrementUserBalance struct {
+	ID  uuid.UUID
+	Sum float32
+}
+
+type WithdrawUserBonuses struct {
+	UserID   uuid.UUID
+	OrderNum string
+	Sum      float32
+}
```
надо в отдельный пакет, можно внутри storage

bulbetski commented on internal/application/storage/storage.go:
```diff
@@ -0,0 +1,55 @@
+	"github.com/google/uuid"
+)
+
+type Storage interface {
```
интерфейсы бы по месту использования

bulbetski commented on internal/auth/jwt.go:
```diff
@@ -0,0 +1,64 @@
+	}
+}
+
+func (j *JWT) GetUserID(_ context.Context, tokenString string) (uuid.UUID, error) {
```
нужен тут контекст?

bulbetski commented on internal/postgres/query.sql:
```diff
@@ -0,0 +1,68 @@
+-- name: GetUser :one
```
Это фигня какая-то, че-то даже не встречал такого. Труднокастомизируемый и трудночитаемый код. 
 В целом наверное поиграться прикольно, но в проде не выживет.

bulbetski commented on internal/postgres/storage.go:
```diff
@@ -0,0 +1,394 @@
+
+type Storage struct {
+	db      *pgxpool.Pool
+	queries *Queries
```
какая-то просто лишняя прослойка получается

bulbetski commented on internal/postgres/storage.go:
```diff
@@ -0,0 +1,394 @@
+	return user, nil
+}
+
+func (s *Storage) WithdrawUserBonuses(ctx context.Context, dto *storage.WithdrawUserBonuses) (*model.User, error) {
```
хорошо, что сделал select for update и использовал транзакцию, но как будто это можно вообще 1 sql запросом сделать

bulbetski reviewed (changes requested):
Оч хорошо, по мелочи отписал

dtroode committed:
refactor: small fixes

dtroode committed:
refactor: medium fixes

dtroode committed:
refactor: move storage interface to place where it is used

dtroode committed:
refactor: Store monetary values as integers

* Refactor: Store monetary values as integers

This commit changes the storage of monetary values (user balances,
order accruals, and withdrawals) from floating-point numbers (`real`)
to integers in the database. All monetary values are now stored as
cents (e.g., $205.60 is stored as 20560).

Key changes include:
- Database migration to change column types to `integer` and convert existing data.
- Updates to `sqlc` schema, queries, and generated models.
- Application models updated to use `int32` for monetary fields internally.
- Storage layer (`internal/postgres/storage.go`) updated to:
    - Convert incoming `float32` DTO values to `int32` cents before DB storage.
    - Map `int32` cents from DB to `int32` in application models.
- Service layer (`internal/application/service/service.go`) updated to:
    - Receive `int32` cents from the storage layer.
    - Convert these `int32` cents to `float32` decimal values for API responses,
      ensuring the external API contract remains unchanged.
- All relevant tests (service layer, job tests) updated to reflect these
  changes, including mock storage interactions and assertions.

This change mitigates potential floating-point precision issues and
ensures more accurate monetary calculations and storage.

* fix: fix some logic

* style: remove comments

---------

Co-authored-by: google-labs-jules[bot] <161369871+google-labs-jules[bot]@users.noreply.github.com>

dtroode committed:
refactor: flatten job

bulbetski commented on internal/application/storage/dto.go:
для storage это уже dbo

bulbetski reviewed (approved):
всё ок
